### PR TITLE
Implement per-thread cache context

### PR DESCRIPTION
**Pull request by beinan**
### What changes are proposed in this pull request?
Implement per-thread cache context

### Why are the changes needed?
Enable the fine-grained cache admission 

### Does this PR introduce any user facing changes?
no

### CODE REVIEW

**Comment by chliang71** [comment]
The change LGTM, thanks for working on this!

**Review by ChunxuTang — APPROVED** [review]
LGTM.
@beinan I saw some testing failures of the `LocalCacheFileInStreamTest` class. Are the failures related to this change?

**Comment by JiamingMai** [comment]
It seems that the test `LocalCacheFileInStreamTest` need to be fixed.

**Comment by beinan** [comment]
alluxio-bot, merge this please.

File: dora/core/client/fs/src/test/java/alluxio/client/file/cache/LocalCacheFileInStreamTest.java [hunk 1]
```diff
@@ -608,6 +610,48 @@ public void testPositionReadFallBack() throws Exception
...
+    // read first page with cache disabled (cache miss)
+    CachePerThreadContext.get().setCacheEnabled(false);
+    stream.positionedRead(0, new byte[mPageSize], 0, mPageSize);
+    Assert.assertEquals(0,
+        MetricsSystem.meter(MetricKey.CLIENT_CACHE_BYTES_READ_CACHE.getName()).getCount());
+    Assert.assertEquals(1,
+        MetricsSystem.counter(MetricKey.CLIENT_CACHE_EXTERNAL_REQUESTS.getName()).getCount());
+    Assert.assertEquals(0,
+        MetricsSystem.counter(MetricKey.CLIENT_CACHE_HIT_REQUESTS.getName()).getCount());
+    // read last page again with cache enabled (cache miss)
```
Comment (robinbach) [inline dora/core/client/fs/src/test/java/alluxio/client/file/cache/LocalCacheFileInStreamTest.java:647]: "read first page again with cache disabled" ?

**Pull request merged**